### PR TITLE
[shelly] Always try to stop WebSocketClient

### PR DESCRIPTION
**Pull request by wborn**
If the client is not stopped it will not be garbage collected because the ShutdownThread keeps a reference to the client so it can be stopped on shutdowns.

See: https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-websocket/websocket-client/src/main/java/org/eclipse/jetty/websocket/client/WebSocketClient.java#L420-L433

### CODE REVIEW

**Review by lolodomo — APPROVED** [review]

**Comment by lolodomo** [comment]
Even if this fix makes sense for us, we will wait for @markus7017 approval this time ;)

**Comment by lolodomo** [comment]
If I correctly remember, few years ago, we fixed all bindings that were stopping the client conditionally. I directly contributed to fix that. Just to say that it was already discussed in the past and it was at this time already a problem for few bindings.

**Comment by lolodomo** [comment]
Look at #8027

**Comment by lolodomo** [comment]
We should not have reverted the fix ;)

**Comment by wborn** [comment]
> Look at https://github.com/openhab/openhab-addons/issues/8027 .

I had forgotten about that one. :+1: 
The mean reason I first wanted a review is that I cannot test the changes because I don't use this binding myself.

**Comment by lolodomo** [comment]
> The main reason I first wanted a review is that I cannot test the changes because I don't use this binding myself.

Yes, that is reasonable.

But at the same time that is a pity to not have this fix in 4.1M2.

**Comment by jlaur** [comment]
I only submitted the PR to revert because of this comment:
https://community.openhab.org/t/oh4-runs-out-of-memory/148699/52

As maintainer I still agree with the change, but as @wborn I don't use the binding myself and cannot assist in testing.

**Review by jlaur — APPROVED** [review]

**Comment by markus7017** [comment]
I agree if the calls to disconnect and close don't through an exception when not connected / open
@wborn Do you know this? Then we could merge

Most likely the binding tries to establish the web socket, which fails and then the socket is not closed, because it was not connected, which then leads into a resource leak

**Comment by wborn** [comment]
From what I can see it won't throw additional exceptions with these changes.

**Comment by jlaur** [comment]
> I agree if the calls to disconnect and close don't through an exception when not connected / open
> @wborn Do you know this? Then we could merge

`disconnect` cannot throw as it catches all exceptions. This includes exceptions from `stop` (which you probably meant to refer to rather than `close`). The only thing that might change is the debug logging. See:

https://github.com/openhab/openhab-addons/blob/f1ccd728a6ef3874410db32f06a08813ae533faf/bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api2/Shelly2RpcSocket.java#L199-L224

So I'll merge.

**Comment by wborn** [comment]
I've now cherry picked this.

**Pull request merged**
